src/RBControl_linesensor.cpp

```diff
@@ -0,0 +1,93 @@
+#include "RBControl_linesensor.hpp"
+
+namespace rb {
+
+
+LineSensor::LineSensor() {
+
+}
+
+LineSensor::~LineSensor() {
+
+}
+
+esp_err_t LineSensor::install(const LineSensor::Config& cfg) {
+    bool expected = false;
+    if(!m_installed.compare_exchange_strong(expected, true))
+        return ESP_OK;
+    
+    esp_err_t ret;
+    spi_bus_config_t buscfg = { 0 };
+    buscfg.miso_io_num = cfg.pin_miso;
+    buscfg.mosi_io_num = cfg.pin_mosi;
+    buscfg.sclk_io_num = cfg.pin_sck;
+    buscfg.quadwp_io_num = -1;
+    buscfg.quadhd_io_num = -1;
+    
+    spi_device_interface_config_t devcfg = { 0 };
+    devcfg.clock_speed_hz = cfg.freq;
+    devcfg.mode = 0;
+    devcfg.spics_io_num = cfg.pin_cs;
+    devcfg.queue_size = LEDS;
+
+    ret = spi_bus_initialize(cfg.spi_dev, &buscfg, 1);
+    if(ret != ESP_OK) {
+        m_installed = false;
+        return ret;
+    }
+
+    ret = spi_bus_add_device(cfg.spi_dev, &devcfg, &m_spi);
+    if(ret != ESP_OK) {
+        spi_bus_free(cfg.spi_dev);
+        m_installed = false;
+        return ret;
+    }
+    return ESP_OK;
+}
+
+esp_err_t LineSensor::read(std::vector<uint16_t>& results, uint8_t leds_mask) {
+    if(!m_installed) {
+        return ESP_FAIL;
+    }
+
+    const size_t orig_size = results.size();
+    int requested = 0;
+
+    spi_transaction_t transactions[LEDS] = { 0 };
```

Comment:
Do you need an array of transactions? Isn't enough to use one instance of `spi_transaction_t` and just rewrite it?

Reply:
Nope, when using the interrupts interface of the spi driver, the spi_transaction_t has to live until you read the result with `spi_device_get_trans_result`